### ISSUE TITLE
Apply requirements seem wrong at worst and confusing at best

### ISSUE THREAD

**Issue by cscalfani** [fantasyland/fantasy-land#142]
Currently, Apply is defined to have the following compositional requirement:

```a.map(f => g => x => f(g(x))).ap(u).ap(v) is equivalent to a.ap(u.ap(v)) (composition)```

This seems wrong. `f`, `g`, and `x` are not found on the righthand side. Neither is `map`.

What am I missing?

**Comment by Avaq** [comment]
`f => g => x => f(g(x))` is `compose`. `f` was supplied by `map`, and is the function contained in `a`. `g` is supplied by `ap` and is the function contained in `u`. `x` is supplied by `ap` and is the value contained in `v`.

So we're using `a f`, `u g` and `v x`, where `a`, `u` and `g` are Apply's and `f` and `g` are functions. On the left hand side we `map(compose).ap(u).ap(v)`, and on the right hand side we manually apply the functions one after the other: `a.ap(u.ap(v))`. It's the "wrapped" equivalent of `compose(f)(g)(x) === f(g(x))`.

**Comment by cscalfani** [comment]
That helps. Thanks.

Would be nice to have annotations in the docs or possibly examples. I looked at the implementation of `Identity` in `ramda-fantasy` to help me understand this better.

The key to my understanding was the fact that `a` contained `f` and `u` contained `g` and `v` contained `x`. This was not intuitive.

BTW, I believe there is a typo in you above explanation. Should read:

[So we're using `a f`, `u g` and `v x`, where `a`, `u` and `v` are Apply's and `f` and `g` are functions.]

I changed the list of applies. It had `g` instead of `v`.

The other typo should read:

[On the left hand side we have `a.map(compose).ap(u).ap(v)`]

I add `a.` in front of the `map` function.



Thanks again for explaining this.
